### PR TITLE
ledger => 3.2.1

### PR DESCRIPTION
**Pull request by supechicken**
**build failed:**
```shell
/usr/local/tmp/crew/ledger.20211105161548.dir/src/format.h:154:35: error: ‘elements’ was not declared in this scope; did you mean ‘element_t’?
  154 |     for (const element_t * elem = elements.get();
      |                                   ^~~~~~~~
      |                                   element_t
/usr/local/tmp/crew/ledger.20211105161548.dir/src/format.
```

### CODE REVIEW

File: packages/ledger.rb [hunk 1]
```diff
@@ -9,19 +9,6 @@ class Ledger < Package
   source_url 'https://github.com/ledger/ledger/archive/v3.1.3.tar.gz'
```
Comment (uberhacker) [inline packages/ledger.rb:9]: Why not update too?  https://github.com/ledger/ledger/archive/refs/tags/v3.2.1.tar.gz

Comment (saltedcoffii) [reply packages/ledger.rb:9]: Or even better, why not just use a git tag instead of an archive

Comment (uberhacker) [reply packages/ledger.rb]: @supechicken: This PR is still in Draft mode.  Were you still testing?

**Comment by uberhacker** [comment]
I just tried the build and got the following:
```
[  4%] Building CXX object src/CMakeFiles/libledger.dir/generate.cc.o
In file included from /usr/local/tmp/crew/ledger.20211106060148.dir/src/report.h:53,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/stats.cc:38:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h:68:5: error: ‘scoped_ptr’ does not name a type; did you mean ‘scope_t’?
   68 |     scoped_ptr<struct element_t> next;
      |     ^~~~~~~~~~
      |     scope_t
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h:106:3: error: ‘scoped_ptr’ does not name a type; did you mean ‘scope_t’?
  106 |   scoped_ptr<element_t> elements;
      |   ^~~~~~~~~~
      |   scope_t
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h: In member function ‘void ledger::format_t::parse_format(const string&, const boost::optional<ledger::format_t&>&)’:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h:138:5: error: ‘elements’ was not declared in this scope; did you mean ‘element_t’?
  138 |     elements.reset(parse_elements(_format, tmpl));
      |     ^~~~~~~~
      |     element_t
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h: In member function ‘virtual void ledger::format_t::mark_uncompiled()’:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h:143:29: error: ‘elements’ was not declared in this scope; did you mean ‘element_t’?
  143 |     for (element_t * elem = elements.get(); elem; elem = elem->next.get()) {
      |                             ^~~~~~~~
      |                             element_t
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h:143:64: error: ‘struct ledger::format_t::element_t’ has no member named ‘next’
  143 |     for (element_t * elem = elements.get(); elem; elem = elem->next.get()) {
      |                                                                ^~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h: In member function ‘virtual void ledger::format_t::dump(std::ostream&) const’:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h:154:35: error: ‘elements’ was not declared in this scope; did you mean ‘element_t’?
  154 |     for (const element_t * elem = elements.get();
      |                                   ^~~~~~~~
      |                                   element_t
/usr/local/tmp/crew/ledger.20211106060148.dir/src/format.h:156:23: error: ‘const struct ledger::format_t::element_t’ has no member named ‘next’
  156 |          elem = elem->next.get())
      |                       ^~~~
In file included from /usr/local/tmp/crew/ledger.20211106060148.dir/src/stats.cc:37:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h: In instantiation of ‘void boost::optional_detail::optional_base<T>::assign_value(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = ledger::account_t::xdata_t; boost::optional_detail::optional_base<T>::rval_reference_type = ledger::account_t::xdata_t&&]’:
/usr/local/include/boost/optional/optional.hpp:350:24:   required from ‘void boost::optional_detail::optional_base<T>::assign(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = ledger::account_t::xdata_t; boost::optional_detail::optional_base<T>::rval_reference_type = ledger::account_t::xdata_t&&]’
/usr/local/include/boost/optional/optional.hpp:1071:21:   required from ‘typename boost::enable_if<boost::is_same<T, typename boost::decay<T_>::type>, boost::optional<T>&>::type boost::optional<T>::operator=(T_&&) [with T_ = ledger::account_t::xdata_t; T = ledger::account_t::xdata_t; typename boost::enable_if<boost::is_same<T, typename boost::decay<T_>::type>, boost::optional<T>&>::type = boost::optional<ledger::account_t::xdata_t>&; typename boost::decay<T_>::type = ledger::account_t::xdata_t]’
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:273:24:   required from here
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:159:10: warning: implicitly-declared ‘ledger::account_t::xdata_t::details_t& ledger::account_t::xdata_t::details_t::operator=(const ledger::account_t::xdata_t::details_t&)’ is deprecated [-Wdeprecated-copy]
  159 |   struct xdata_t : public supports_flags<>
      |          ^~~~~~~
In file included from /usr/local/tmp/crew/ledger.20211106060148.dir/src/iterators.h:47,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/generate.h:45,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/generate.cc:34:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h: In instantiation of ‘void boost::optional_detail::optional_base<T>::assign_value(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = ledger::account_t::xdata_t; boost::optional_detail::optional_base<T>::rval_reference_type = ledger::account_t::xdata_t&&]’:
/usr/local/include/boost/optional/optional.hpp:350:24:   required from ‘void boost::optional_detail::optional_base<T>::assign(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = ledger::account_t::xdata_t; boost::optional_detail::optional_base<T>::rval_reference_type = ledger::account_t::xdata_t&&]’
/usr/local/include/boost/optional/optional.hpp:1071:21:   required from ‘typename boost::enable_if<boost::is_same<T, typename boost::decay<T_>::type>, boost::optional<T>&>::type boost::optional<T>::operator=(T_&&) [with T_ = ledger::account_t::xdata_t; T = ledger::account_t::xdata_t; typename boost::enable_if<boost::is_same<T, typename boost::decay<T_>::type>, boost::optional<T>&>::type = boost::optional<ledger::account_t::xdata_t>&; typename boost::decay<T_>::type = ledger::account_t::xdata_t]’
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:273:24:   required from here
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:159:10: warning: implicitly-declared ‘ledger::account_t::xdata_t::details_t& ledger::account_t::xdata_t::details_t::operator=(const ledger::account_t::xdata_t::details_t&)’ is deprecated [-Wdeprecated-copy]
  159 |   struct xdata_t : public supports_flags<>
      |          ^~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:215:7: note: because ‘ledger::account_t::xdata_t::details_t’ has user-provided ‘ledger::account_t::xdata_t::details_t::details_t(const ledger::account_t::xdata_t::details_t&)’
  215 |       details_t(const details_t&)
      |       ^~~~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:215:7: note: because ‘ledger::account_t::xdata_t::details_t’ has user-provided ‘ledger::account_t::xdata_t::details_t::details_t(const ledger::account_t::xdata_t::details_t&)’
  215 |       details_t(const details_t&)
      |       ^~~~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:159:10: warning: implicitly-declared ‘ledger::account_t::xdata_t::details_t& ledger::account_t::xdata_t::details_t::operator=(const ledger::account_t::xdata_t::details_t&)’ is deprecated [-Wdeprecated-copy]
  159 |   struct xdata_t : public supports_flags<>
      |          ^~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:215:7: note: because ‘ledger::account_t::xdata_t::details_t’ has user-provided ‘ledger::account_t::xdata_t::details_t::details_t(const ledger::account_t::xdata_t::details_t&)’
  215 |       details_t(const details_t&)
      |       ^~~~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:159:10: warning: implicitly-declared ‘ledger::account_t::xdata_t::details_t& ledger::account_t::xdata_t::details_t::operator=(const ledger::account_t::xdata_t::details_t&)’ is deprecated [-Wdeprecated-copy]
  159 |   struct xdata_t : public supports_flags<>
      |          ^~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/account.h:215:7: note: because ‘ledger::account_t::xdata_t::details_t’ has user-provided ‘ledger::account_t::xdata_t::details_t::details_t(const ledger::account_t::xdata_t::details_t&)’
  215 |       details_t(const details_t&)
      |       ^~~~~~~~~
In file included from /usr/local/include/boost/optional.hpp:15,
                 from /usr/local/include/boost/format/internals.hpp:21,
                 from /usr/local/include/boost/format.hpp:38,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/system.hh:152,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/CMakeFiles/libledger.dir/cmake_pch.hxx:5,
                 from <command-line>:
/usr/local/include/boost/optional/optional.hpp:751:64: note: synthesized method ‘ledger::account_t::xdata_t& ledger::account_t::xdata_t::operator=(const ledger::account_t::xdata_t&)’ first required here
  751 |     void assign_value ( rval_reference_type val ) { get_impl() = static_cast<rval_reference_type>(val); }
      |                                                     ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/include/boost/optional.hpp:15,
                 from /usr/local/include/boost/format/internals.hpp:21,
                 from /usr/local/include/boost/format.hpp:38,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/system.hh:152,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/CMakeFiles/libledger.dir/cmake_pch.hxx:5,
                 from <command-line>:
/usr/local/include/boost/optional/optional.hpp:751:64: note: synthesized method ‘ledger::account_t::xdata_t& ledger::account_t::xdata_t::operator=(const ledger::account_t::xdata_t&)’ first required here
  751 |     void assign_value ( rval_reference_type val ) { get_impl() = static_cast<rval_reference_type>(val); }
      |                                                     ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/tmp/crew/ledger.20211106060148.dir/src/report.h:46,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/stats.cc:38:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/query.h: In instantiation of ‘void boost::optional_detail::optional_base<T>::assign_value(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = ledger::query_t::parser_t; boost::optional_detail::optional_base<T>::rval_reference_type = ledger::query_t::parser_t&&]’:
/usr/local/include/boost/optional/optional.hpp:350:24:   required from ‘void boost::optional_detail::optional_base<T>::assign(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = ledger::query_t::parser_t; boost::optional_detail::optional_base<T>::rval_reference_type = ledger::query_t::parser_t&&]’
/usr/local/include/boost/optional/optional.hpp:1071:21:   required from ‘typename boost::enable_if<boost::is_same<T, typename boost::decay<T_>::type>, boost::optional<T>&>::type boost::optional<T>::operator=(T_&&) [with T_ = ledger::query_t::parser_t; T = ledger::query_t::parser_t; typename boost::enable_if<boost::is_same<T, typename boost::decay<T_>::type>, boost::optional<T>&>::type = boost::optional<ledger::query_t::parser_t>&; typename boost::decay<T_>::type = ledger::query_t::parser_t]’
/usr/local/tmp/crew/ledger.20211106060148.dir/src/query.h:330:58:   required from here
/usr/local/tmp/crew/ledger.20211106060148.dir/src/query.h:249:9: warning: implicitly-declared ‘ledger::query_t::lexer_t& ledger::query_t::lexer_t::operator=(const ledger::query_t::lexer_t&)’ is deprecated [-Wdeprecated-copy]
  249 |   class parser_t
      |         ^~~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/query.h:213:5: note: because ‘ledger::query_t::lexer_t’ has user-provided ‘ledger::query_t::lexer_t::lexer_t(const ledger::query_t::lexer_t&)’
  213 |     lexer_t(const lexer_t& lexer)
      |     ^~~~~~~
/usr/local/tmp/crew/ledger.20211106060148.dir/src/query.h:249:9: warning: implicitly-declared ‘ledger::keep_details_t& ledger::keep_details_t::operator=(const ledger::keep_details_t&)’ is deprecated [-Wdeprecated-copy]
  249 |   class parser_t
      |         ^~~~~~~~
In file included from /usr/local/tmp/crew/ledger.20211106060148.dir/src/predicate.h:47,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/xact.h:46,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/stats.cc:35:
/usr/local/tmp/crew/ledger.20211106060148.dir/src/annotate.h:132:3: note: because ‘ledger::keep_details_t’ has user-provided ‘ledger::keep_details_t::keep_details_t(const ledger::keep_details_t&)’
  132 |   keep_details_t(const keep_details_t& other)
      |   ^~~~~~~~~~~~~~
In file included from /usr/local/include/boost/optional.hpp:15,
                 from /usr/local/include/boost/format/internals.hpp:21,
                 from /usr/local/include/boost/format.hpp:38,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/system.hh:152,
                 from /usr/local/tmp/crew/ledger.20211106060148.dir/src/CMakeFiles/libledger.dir/cmake_pch.hxx:5,
                 from <command-line>:
/usr/local/include/boost/optional/optional.hpp:751:64: note: synthesized method ‘ledger::query_t::parser_t& ledger::query_t::parser_t::operator=(const ledger::query_t::parser_t&)’ first required here
  751 |     void assign_value ( rval_reference_type val ) { get_impl() = static_cast<rval_reference_type>(val); }
      |                                                     ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [src/CMakeFiles/libledger.dir/build.make:93: src/CMakeFiles/libledger.dir/stats.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:235: src/CMakeFiles/libledger.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
ledger failed to build: `make V=0 -j2` exited with 2
```

**Comment by uberhacker** [comment]
Let's revisit this with a new PR that we know works.   We really shouldn't be submitting WIP PRs anyway.

**Comment by satmandu** [comment]
How about we have a general rule that if you're submitting a PR you have tested that it builds in a fresh container, or at least locally.

Having said that, it's entirely possible that it builds locally but for some reason won't build for someone else.

**Comment by satmandu** [comment]
(maybe that means we ought to look at getting docker working too on chromebooks...)

**Pull request closed without merging**